lib/internal/streams/operators.js

```diff
@@ -0,0 +1,234 @@
+'use strict';
+
+const { AbortController } = require('internal/abort_controller');
+const Readable = require('internal/streams/readable');
+const { isNodeStream, isReadableNodeStream, isIterable } = require('internal/streams/utils');
+const { addAbortSignal } = require('internal/streams/add-abort-signal');
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+const console = require('console');
+
+const kEmpty = Symbol('kEmpty');
+
+async function * mapImpl(stream, fn, options, config) {
+  if (typeof fn !== 'function') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'fn', ['Function', 'AsyncFunction'], stream);
+  }
+
+  let concurrency = 1;
+  if (options && options.concurrency != null) {
+    concurrency = options.concurrency;
+  } else if (options != null && typeof options !== 'object') {
+    concurrency = options;
+  }
+
+  if (!Number.isInteger(concurrency)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'concurrency', ['Integer'], concurrency);
+  }
+
+  if (concurrency < 1) {
+    throw new ERR_INVALID_ARG_VALUE('concurrency', concurrency);
+  }
+
+  const ac = new AbortController();
+  const abort = () => ac.abort();
+
+  const queue = [];
+  const signalOpt = { signal: ac.signal };
+  const flatten = !!config?.flatten;
+  const outerSignal = options?.signal;
+
+  outerSignal?.addEventListener('abort', abort);
+
+  let next;
+  let resume;
+  let done = false;
+
+  async function pump () {
+    try {
+      for await (let val of stream) {
+        if (done) {
+          return;
+        }
+
+        try {
+          val = fn(val, signalOpt);
+        } catch (err) {
+          val = Promise.reject(err);
+        }
+
+        if (val === kEmpty) {
+          continue;
+        }
+
+        if (typeof val?.then === 'function') {
+          val.then(null, () => {
+            done = true;
+          });
+        }
+
+        queue.push(val);
+        if (next) {
+          next();
+          next = null;
+        }
+
+        if (!done && queue.length >= concurrency) {
+          await new Promise(resolve => {
+            resume = resolve;
+          });
+        }
+      }
+      queue.push(null);
+    } catch (err) {
+      queue.push(Promise.reject(err));
```

Comment:
Does this need a .catch(noop) or will it immediately be consumed in this case?

Reply:
you are right